site-src/geps/gep-1323.md

```diff
@@ -0,0 +1,204 @@
+* Issue: [#1323](https://github.com/kubernetes-sigs/gateway-api/issues/1323)
+* Status: Implementable
+
+## TLDR
+Similar to how we have `RequestHeaderModifier` in `HTTPRouteFilter`, which lets users modify request headers before the request is forwarded to a backend (or a group of backends), it’d be helpful to have a `ResponseHeaderModifier` field which would let users modify response headers before they are forwarded to its destination.
+
+## Goals
+* Provide a way to modify HTTP response headers.
+* Reuse existing types as much as possible to reduce boilerplate code.
+
+## Non Goals
+* Provide a way to modify other parts of a HTTP response like status code.
+* Add fields specifically for standard headers such as `Cookie`.
+
+## Introduction
+Currently, the `HTTPRouteFilter` API provides a way for request headers to be modified through the `RequestHeaderModifier` field of type `HTTPRequestHeaderModifier`. But, a similar API to modify response headers does not exist. This proposal intends to introduce a new field in `HTTPRouteFilter` named `ResponseHeaderModifier`.
+
+## API
+We could introduce a new API named `HTTPResponseHeaderModifier` which would look exactly like the existing `HTTPRequestHeaderModifier` API. But since HTTP headers have the same semantics for both requests and responses, it makes more sense to rename `HTTPRequestHeaderModifier` to `HTTPHeaderModifier` and use this for both `RequestHeaderModifier` and `ResponseHeaderModifier`.
+
+```golang
+// HTTPHeaderModifier defines a filter that modifies the headers of a HTTP
+// request or response.
+type HTTPHeaderModifier struct {
+    // Set overwrites the request with the given header (name, value)
+    // before the action.
+    // +optional
+    // +listType=map
+    // +listMapKey=name
+    // +kubebuilder:validation:MaxItems=16
+    Set []HTTPHeader `json:"set,omitempty"`
+
+    // Add adds the given header(s) (name, value) to the request
+    // before the action. It appends to any existing values associated
+    // with the header name.
+
+    // +optional
+    // +listType=map
+    // +listMapKey=name
+    // +kubebuilder:validation:MaxItems=16
+    Add []HTTPHeader `json:"add,omitempty"`
+
+    // Remove the given header(s) from the HTTP request before the action. The
+    // value of Remove is a list of HTTP header names. Note that the header
+    // names are case-insensitive (see
+    // https://datatracker.ietf.org/doc/html/rfc2616#section-4.2).
+    // +optional
+    // +kubebuilder:validation:MaxItems=16
+    Remove []string `json:"remove,omitempty"`
+}
+```
+
+Given the fact that this functionality is offered by only a few projects that currently implement Gateway API when using their own traffic routing CRDs, it’s better to support `ResponseHeaderModifier` as an _Extended_ feature, unlike `RequestHeaderModifier` which is a _Core_ feature. This will also not increase the difficulty of implementing Gateway API for any future ingress or service mesh.
```

Comment:
I like this, we can use the same `HTTPHeaderModifier` struct to represent the actual config, but using the field that the struct appears in inside the Route can have different support, and that support can be documented on those fields.

Not sure if we need that in this GEP or not though.

Reply:
Do you mean we could add a line in the `HTTPRouteFilter` API above the `ResponseHeaderModifier` field mentioning the support similar to this: https://github.com/kubernetes-sigs/gateway-api/blob/bb549d06152f11759358298970a807e8ffb08089/apis/v1beta1/httproute_types.go#L571?
If yes, I didn't mention it in the proposal, because I thought it's implied that the support will be documented in the API docs. But if you think, we need to be explicit about it in the proposal, I'll add something accordingly.

Reply:
Yes, I think that when this is implemented, any new fields should have their support explicitly documented.